src/engine/effects/engineeffectrack.cpp

```diff
@@ -41,16 +44,52 @@ bool EngineEffectRack::processEffectsRequest(const EffectsRequest& message,
     return true;
 }
 
-void EngineEffectRack::process(const ChannelHandle& handle,
-                               CSAMPLE* pInOut,
+bool EngineEffectRack::process(const ChannelHandle& inputHandle,
+                               const ChannelHandle& outputHandle,
+                               CSAMPLE* pIn, CSAMPLE* pOut,
                                const unsigned int numSamples,
                                const unsigned int sampleRate,
                                const GroupFeatureState& groupFeatures) {
-    foreach (EngineEffectChain* pChain, m_chains) {
-        if (pChain != NULL) {
-            pChain->process(handle, pInOut, numSamples, sampleRate, groupFeatures);
+    int chainsProcessed = 0;
+    if (pIn == pOut) {
+        // Effects are applied to the buffer in place
+        for (EngineEffectChain* pChain : m_chains) {
+            if (pChain != nullptr) {
+                if (pChain->process(inputHandle, outputHandle,
+                                    pIn, pOut,
+                                    numSamples, sampleRate, groupFeatures)) {
+                    ++chainsProcessed;
+                }
+            }
+        }
+    } else {
+        // Do not modify the input buffer; only fill the output buffer.
+        CSAMPLE* pIntermediateInput = pIn;
+        CSAMPLE* pIntermediateOutput = m_buffer1.data();
+
+        for (EngineEffectChain* pChain : m_chains) {
+            if (pChain != nullptr) {
+                if (pChain->process(inputHandle, outputHandle,
+                                    pIntermediateInput, pIntermediateOutput,
+                                    numSamples, sampleRate, groupFeatures)) {
+                    ++chainsProcessed;
+                    if (chainsProcessed % 2) {
```

Comment:
Very similar code as in EngineEffectsManager, the same comments apply.

We wouldn't need to duplicate code if the processing units of the engine would be more versatile and modular. Sometimes in the far future this should become a modular design. Unfortunately there is no open standard (VST2/3?) and AUs are tied to macOS and iOS. In general I like their approach even though the API looks complex.

Reply:
"The" open audio plugin interface for is called LV2: http://lv2plug.in/ Apart from VAMP for analysis, (and LADSPA if you're interested in legacy plugins..) LV2 is the way to go for plugins.

Starting by abstracting the various functionality into a single class is a good first step either way though, so I'd do that first before digging into LV2 (which is very nicely designed from a technical POV, but can be a bit verbose when starting).

Reply:
We have a branch for LV2 support that was started by a GSOC student a few years ago: #1240. However, it will require quite a bit of work adding capabilities to the effects system to make it usable, such as user-defined parameter rearrangement and hiding.

Reply:
When I looked at the code of an LV2 effect, I realized our effects system's states may be overcomplicated. LV2 has a single method that is called when an effect is enabled. We have separate states for enabling and disabling. We may be able to simplify this so everything that is done at disabling is done at the start of enabling.